core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/StitchAppRequestClient.java

```diff
@@ -0,0 +1,180 @@
+import com.mongodb.stitch.core.StitchRequestErrorCode;
+import com.mongodb.stitch.core.StitchRequestException;
+import com.mongodb.stitch.core.internal.common.IoUtils;
+import com.mongodb.stitch.core.internal.common.StitchObjectMapper;
+
+import java.io.ByteArrayOutputStream;
+import java.nio.charset.StandardCharsets;
+
+public class StitchAppRequestClient extends StitchRequestClient {
+  private static final String BOOTSTRAP_ERROR_MESSAGE_INVALID_HOSTNAME =
+      "invalid hostname in metadata: %s";
+
+  private final String clientAppId;
+
+  private AppMetadata appMetadata;
+
+  public StitchAppRequestClient(
+      final String clientAppId,
+      final String baseUrl,
+      final Transport transport,
+      final Long defaultRequestTimeout
+  ) {
+    super(baseUrl, transport, defaultRequestTimeout);
+    this.clientAppId = clientAppId;
+  }
+
+  /**
+   * Performs a request against a Stitch app server determined by the deployment model
+   * of the underlying app. Throws a Stitch specific exception if the request fails.
+   *
+   * @param stitchReq the request to perform.
+   * @return a {@link Response} to the request.
+   */
+  @Override
+  public Response doRequest(final StitchRequest stitchReq) {
+    if (appMetadata == null) {
+      init(clientAppId);
+    }
+    // init throws if no location is available, so this should be safe
+
+    final Response response;
+    try {
+      response = transport.roundTrip(buildRequest(stitchReq, appMetadata.hostname));
+    } catch (Exception e) {
+      throw new StitchRequestException(e, StitchRequestErrorCode.TRANSPORT_ERROR);
+    }
+
+    inspectResponse(response);
+
+    return response;
+  }
+
+  /**
+   * Performs a streaming request against a Stitch app server determined by the deployment model
+   * of the underlying app. Throws a Stitch specific exception if the request fails.
+   *
+   * @param stitchReq the request to perform.
+   * @return a {@link EventStream} representing the result of the request.
+   */
+  @Override
+  public EventStream doStreamRequest(final StitchRequest stitchReq) {
+    if (appMetadata == null) {
+      init(clientAppId);
+    }
+    // init throws if no location is available, so this should be safe
+
+    try {
+      return transport.stream(buildRequest(stitchReq, appMetadata.hostname));
+    } catch (Exception e) {
+      throw new StitchRequestException(e, StitchRequestErrorCode.TRANSPORT_ERROR);
+    }
+  }
+
+  private synchronized void init(final String clientAppId) {
+    // double-check, in case two threads hit this point simultaneously
+    if (appMetadata != null) {
+      return;
+    }
+
+    final StitchAppRoutes routes = new StitchAppRoutes(clientAppId);
+
+    final StitchRequest bootstrapStitchRequest = new StitchRequest.Builder()
+        .withMethod(Method.GET)
+        .withPath(routes.getServiceRoutes().getLocationRoute())
+        .build();
+    final Request bootstrapRequest = buildRequest(bootstrapStitchRequest, baseUrl);
```

Comment:
[minor] For the code here and below, you should just be able to reuse super's `doRequest` and then decode it like you are at the bottom here. That would avoid some duplicate code.

Reply:
done